src/ray/util/function_traits.h

```diff
@@ -0,0 +1,68 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <boost/type_traits/function_traits.hpp>
+#include <functional>
+#include <type_traits>
+
+namespace ray {
+
+namespace internal {
```

Comment:
btw you need a unit test for the function traits

Something similar to
```
double SimpleFunction(int);

struct Foo {
  double member(int);
  double const_member(int) const;
  double move_member(int) &&; // NOLINT
};

TEST(SignatureTest, Basic) {
  using E = signature_t<double(int)>;
  using CE = signature_t<double(int) const>;
  {
    using T = signature_t<decltype(SimpleFunction)>;
    static_assert(std::is_same_v<T, E>);
  }
  {
    using T = signature_t<decltype(&SimpleFunction)>;
    static_assert(std::is_same_v<T, E>);
  }
  {
    using T = signature_t<decltype(&Foo::member)>;
    static_assert(std::is_same_v<T, E>);
  }
  {
    using T = signature_t<decltype(&Foo::const_member)>;
    static_assert(std::is_same_v<T, CE>);
  }
  {
    using T = signature_t<decltype(&Foo::move_member)>;
    static_assert(std::is_same_v<T, E>);
  }
  {
    using T = signature_t<std::function<double(int)>>;
    static_assert(std::is_same_v<T, CE>);
  }
  {
    using T = signature_t<std::function<double(int)>&>;
    static_assert(std::is_same_v<T, CE>);
  }
  {
    auto plain_lambda = [](int i) { return 2. * i; };
    using T = signature_t<decltype(plain_lambda)>;
    static_assert(std::is_same_v<T, CE>);
  }
  {
    const double f = 2.;
    auto capturing_lambda = [f](int i) { return f * i; };
    using T = signature_t<decltype(capturing_lambda)>;
    static_assert(std::is_same_v<T, CE>);
  }
  {
    auto f = std::make_unique<double>(2.);
    auto moveonly_lambda = [f = std::move(f)](int i) { return *f * i; };
    using T = signature_t<decltype(moveonly_lambda)>;
    static_assert(std::is_same_v<T, CE>);
  }
}
```

Reply:
thanks for the unit test! It actually caught an issue: in matching std::function I got an ambiguous template error and fixed.